hydra-node/src/Hydra/Chain/Direct/Tx.hs

```diff
@@ -217,26 +242,82 @@ commitTx ::
   ScriptRegistry ->
   HeadId ->
   Party ->
-  -- | The UTxO to commit to the Head along with witnesses.
-  UTxO' (TxOut CtxUTxO, Witness WitCtxTxIn) ->
+  -- | The UTxO to commit to the Head.
+  UTxO ->
+  -- | _Blueprint_ transaction.
+  Tx ->
   -- | The initial output (sent to each party) which should contain the PT and is
   -- locked by initial script
   (TxIn, TxOut CtxUTxO, Hash PaymentKey) ->
   Tx
-commitTx networkId scriptRegistry headId party utxoToCommitWitnessed (initialInput, out, vkh) =
-  unsafeBuildTransaction $
-    emptyTxBody
-      & addInputs [(initialInput, initialWitness)]
-      & addReferenceInputs [initialScriptRef]
-      & addInputs committedTxIns
-      & addExtraRequiredSigners [vkh]
-      & addOutputs [commitOutput]
-      & setTxMetadata (TxMetadataInEra $ mkHydraHeadV1TxName "CommitTx")
+commitTx networkId scriptRegistry headId party utxoToCommit blueprintTx (initialInput, out, vkh) =
+  let
+    ledgerBlueprintTx =
+      toLedgerTx blueprintTx
+        & bodyTxL . inputsTxBodyL <>~ Set.singleton (toLedgerTxIn initialInput)
+        & bodyTxL . inputsTxBodyL <>~ Set.fromList (toLedgerTxIn <$> committedTxIns)
+        & bodyTxL . referenceInputsTxBodyL <>~ Set.fromList [toLedgerTxIn initialScriptRef]
+        & bodyTxL . outputsTxBodyL %~ (toLedgerTxOut commitOutput <|)
+        & bodyTxL . reqSignerHashesTxBodyL <>~ Set.singleton (toLedgerKeyHash vkh)
+        & bodyTxL . auxDataHashTxBodyL .~ SJust (hashAlonzoTxAuxData txAuxMetadata)
+        & auxDataTxL .~ SJust txAuxMetadata
+    existingWits = ledgerBlueprintTx ^. witsTxL
+    currentInputs = ledgerBlueprintTx ^. bodyTxL . inputsTxBodyL
+    commitInputs = UTxO.inputSet utxoToCommit
+    blueprintRedeemer = unRedeemers $ toLedgerTx blueprintTx ^. witsTxL . rdmrsTxWitsL
+    reAssociatedTxIns = reAssociateOldInputs currentInputs commitInputs blueprintRedeemer
+    wits =
+      witsTxL
+        .~ ( existingWits
+              & rdmrsTxWitsL .~ rebuildRedeemers reAssociatedTxIns (initialInput, initialRedeemer)
+           )
```

Comment:
Should: simplify this rebuilding / reassociation

This looks very fragile and hard to maintain as it handles indices directly. Not a must-have, but I would suggest this algorithm:
1. Resolve redeemers in `blueprintTx` so we know which `TxIn` they spend, e.g. as a `Map (PlutusPurpose AsItem) Data`
2. Add the `initialRedeemer` spending `initialInput` to that map
3. Build the final commitTx body with blueprint tx inputs + `initialInput` 
4. Create the `Redeemers` using `redeemerPointer` or a similar function translating from `TxIn` to the indices.

Reply:
I couldn't see how to resolve the final part of your algorithm and instead simplified redeemer re-building so it is contained in a single smaller function.